### ISSUE TITLE
Uniform magnitude pruning implementation problem

### ISSUE THREAD

**Issue by LYF915** [IntelLabs/Model-Compression-Research-Package#11]
Hello, when the uniform magnitude pruning method is set to
"pruning_fn_default_kwargs": {
     "block_size": 8,
     "target_sparsity": 0.85
   },
The model ends up retaining the parameter 0.75, why?

**Comment by ofirzaf** [comment]
Uniform pruning makes sure that in each block in the size you define will be pruned to be at the target sparsity at the end of training, at the end you will achieve a sparsity pattern of N:M. For example, if you define `block_size=4, target_sparsity=0.5` you will get 2:4 sparsity pattern.

In your case 0.85*8 = 6.8, in the implementation here this number is always rounded down: https://github.com/IntelLabs/Model-Compression-Research-Package/blob/69aecbf5cc73b10fab88a13d8ca6d8314d284c0b/model_compression_research/pruning/methods/methods_utils.py#L34
so you get 6 zeros in every block of 8 weights resulting in 75% sparsity. If you will increase the `target_sparsity` to 0.875 you will get 7 zeros per block which might be closer to what you are trying to do.

If this is not the kind of sparsity pattern you were looking for, let me know what you are aiming at and I will try to point you to the method that will best suit your needs.

**Comment by LYF915** [comment]
I understand the previous question, but I have one more question about uniforms. The config settings of pruning are as follows: 
{
  "begin_pruning_step": 0,
  "end_pruning_step": 1200,
  "not_to_prune": ["classifier", "teacher"],
  "policy_begin_step": 0,
  "policy_end_step": 800,
  "prune_layer_types": {
    "Linear": {"name": "weight"}
  },
  "pruning_fn": "uniform_magnitude",
  "pruning_fn_default_kwargs": {"target_sparsity": 0.75,"block_size":8},
  "pruning_frequency": 100,
  "scheduler": "iterative"
}

But when my task is set to sst2, the final sparsity of the model is 0.25 as expected. But when I set to tasks like qnli, there is an exception in model sparsity.
![截屏2022-06-29 12 07 13](https://user-images.githubusercontent.com/43818899/176349379-b5f40ed4-b0a9-43b4-8779-82ca262fc0b0.png)

I haven't figured out why this phenomenon is happening, can you give me a hint?

**Comment by ofirzaf** [comment]
Which model are you trying to prune and how do you calculate the measurements above? 
In which dimension do you require the uniform pruning pattern?

**Comment by LYF915** [comment]
The model I use is bert, and the parameter is the total number of non-zero parameters of the trimmed encoder divided by the total number of non-zero parameters of the original encoder.

But now I found the problem, I used the load_best_model_at_end & EarlyStop strategy, which resulted in loading the model that was not fully pruned. As a result, an error was caused. I placed a conditional restriction on on_save to ensure that the model after prune be stored.

Is there any documentation describing the various compression methods for this package? 
pattern_lock is to make the pruned model fix the 0 parameter position and fine-tune other non-0 parameters? Is it automatically enabled after end_pruning_step? Can this mode be used alone (it is the pruned model is loaded, use the pattern_lock to continue to fine-tune the pruned model ?)

**Comment by LYF915** [comment]
With your help, I have a general understanding of this package, so this compressed package contains a total of two unstructured pruning methods (threshold pruning, STE self-learning pruning)? Are you considering adding structured pruning content?

In addition, I am curious whether the Qint8 method can be compared with the pruned model with 0.75 sparsity? Can the two be regarded as the same sparsity?

**Comment by ofirzaf** [comment]
The package support uniform and block structured pruning in addition to unsdtructured pruning. There is also support for global pruning (treating all the weights in the model as 1 big weight when computing the pruning threshold).

I am not sure what you mean by "STE self-learning pruning"?

You are welcome to implement other pruning algorithms. We tried to make it as easy as possible to extend the package to other pruning methods aswell. Let me know if you want some help to do that

Quantization method is completely orthogonal to pruning and you can apply both on your model. for example if you prune 75% of the model and also quantize it to int8 you will get up to 16x compression since you get 4x compression from sparsity and an additional 4x compression from quantization. I don't think you can view quantization as a type of sparsity.

**Comment by LYF915** [comment]
STE self-learning pruning here I refer to the neural wiring pruning method in the package.

Although the quantization and pruning methods are technically orthogonal, both are lossy compression in practice. Therefore, if I need to achieve the optimal performance under the guarantee of 4x compression ratio acceleration, can the two methods be compared directly and fairly?

**Comment by LYF915** [comment]
I have used torch to quantize the model to int8 and then in cpu. The speedup of the inference actually is <<4x, and I found that as the number of CPUs increases (>16), the speed of the model after quantization is not improved (the model will continue to accelerate after onnx runtime quantization, but the acceleration effect will slow down), Have you ever come across this situation? What do you think is the reason?

If we just look at the actual inferred speedup ratio, is pruning closer to the theoretical speedup ratio?

**Comment by LYF915** [comment]
> I have used torch to quantize the model to int8 and then in cpu. The speedup of the inference actually is <<4x, and I found that as the number of CPUs increases (>16), the speed of the model after quantization is not improved (the model will continue to accelerate after onnx runtime quantization, but the acceleration effect will slow down), Have you ever come across this situation? What do you think is the reason?
> 
> If we just look at the actual inferred speedup ratio, is pruning closer to the theoretical speedup ratio?

I mean Pareto optimal solution of physical speedup & performance. Which compression method is better, have you read relevant research? Or according to your experience?

Also thank you for your patience, your answer has helped me a lot!

**Comment by ofirzaf** [comment]
From my experience quantization will give you the best performance for 4x compression. Accelerating sparse NN is still very new and not a lot of engines are capable of doing that. AFAIK only NVIDIA (2:4 sparsity using A100 GPU) and Neural Magic accelerate sparse NN.

When running a model in production you should always use a runtime engine for inference such as OpenVino/ONNX-runtime since it will always give you better performance than a Python based runtime such as Pytorch/Tensorflow (from my experience).

If your performance gain gets saturated as you increase the number of cores that means you have become memory bound and the CPU can't bring memory fast enough to the cores and they sit in idle until they can resume computing. A solution to that might be multi-instance where you have a few independent copies of the same model and you feed each copy with different data and compute in parallel.

**Comment by LYF915** [comment]
Thanks, I learned a lot from your answer!
